_posts/2017-11-23-totw-126.md

```diff
@@ -0,0 +1,117 @@
+---
+title: "Tip of the Week #126: make_unique is the new `new`"
+layout: tips
+sidenav: side-nav-tips.html
+published: false
+permalink: tips/126
+type: markdown
+order: "126"
+---
+
+Originally posted as totw/126 on 2016-12-12
+
+By James Dennett (jdennett@google.com) based on a mailing list post by Titus
+Winters (titus@google.com)
+
+As a codebase expands it is increasingly difficult to know the details of
+everything you depend on. Requiring deep knowledge doesn't scale: we have to
+rely on interfaces and contracts to know that code is correct, both when writing
+and when reviewing. In many cases the type system can provide those contracts in
+a common fashion. Consistent use of type system contracts makes for easier
+authoring and reviewing of code by identifying places where there are
+potentially risky allocations or ownership transfers for objects allocated on
+the heap.
+
+While in C++ we can reduce the need for dynamic memory allocation by using plain
+values, sometimes we need objects to outlive their scope. C++ code should prefer
+smart pointers (most commonly `std::unique_ptr`) instead of raw pointers when
+dynamically allocating objects. This provides a consistent story around
+allocation and ownership transfer, and leaves a clearer visual signal where
+there's code that needs closer inspection for ownership issues. The side effect
+of matching how allocation works in the outside world post-C++14 and being
+exception safe is just icing.
+
+Two key tools for this are `absl::make_unique` (a C++11 implementation of
+C++14's `std::make_unique`, for leak-free dynamic allocation) and
+`absl::WrapUnique` (for wrapping owned raw pointers into the corresponding
+`std::unique_ptr` types). They can be found in
+[absl/memory/memory.h](https://github.com/abseil/abseil-cpp/blob/master/absl/memory/memory.h). Why
+should code prefer smart pointers and these `absl` functions over raw pointers
+and `new`?
+
+1.  When possible, ownership is best expressed in the type system. This allows
+    reviewers to verify correctness (absence of leaks and of double-deletes)
+    almost entirely by local inspection. (In code that is exceptionally
+    performance sensitive, this may be excused: while cheap, passing
+    `std::unique_ptr` across function boundaries by value has non-zero overhead
+    because of ABI constraints. That's rarely important enough to justify
+    avoiding it.)
+
+2.  Somewhat like the reasoning for preferring `push_back` over `emplace` ([TotW
```

Comment:
FWIW, I mildly prefer moving while link text to new lines if possible

([TotW 112)](/tips/112)